llvm/lib/Target/DirectX/DXILRootSignature.cpp

```diff
@@ -0,0 +1,215 @@
+      continue;
+    }
+
+    if (F != EntryFunction)
+      continue;
+
+    // Get the Root Signature Description from the function signature pair.
+    MDNode *RS = dyn_cast<MDNode>(Node->getOperand(1).get());
+
+    if (RS == nullptr) {
+      reportError(Ctx, "Missing Root Element List Metadata node.");
+      continue;
+    }
+
+    // Loop through the Root Elements of the root signature.
+    for (const auto &Operand : RS->operands()) {
+      MDNode *Element = dyn_cast<MDNode>(Operand);
+      if (Element == nullptr)
+        return reportError(Ctx, "Missing Root Element Metadata Node.");
+
+      HasError = HasError || parseRootSignatureElement(Ctx, MRS, Element);
+    }
+  }
+  return HasError;
+}
+
+static bool validate(LLVMContext *Ctx, ModuleRootSignature *MRS) {
+  if (!dxbc::RootSignatureValidations::isValidRootFlag(MRS->Flags)) {
+    return reportError(Ctx, "Invalid Root Signature flag value");
+  }
+  return false;
+}
+
+static const Function *getEntryFunction(Module &M, ModuleMetadataInfo MMI) {
+
+  LLVMContext *Ctx = &M.getContext();
+  if (MMI.EntryPropertyVec.size() != 1) {
+    reportError(Ctx, "More than one entry function defined.");
+    // needed to stop compilation
+    report_fatal_error("Invalid Root Signature Definition", false);
+    return nullptr;
+  }
+  return MMI.EntryPropertyVec[0].Entry;
+}
+
+std::optional<ModuleRootSignature>
+ModuleRootSignature::analyzeModule(Module &M, const Function *F) {
+
+  LLVMContext *Ctx = &M.getContext();
+
+  ModuleRootSignature MRS;
+
+  NamedMDNode *RootSignatureNode = M.getNamedMetadata("dx.rootsignatures");
+  if (RootSignatureNode == nullptr)
+    return std::nullopt;
+
+  if (parse(Ctx, &MRS, RootSignatureNode, F) || validate(Ctx, &MRS)) {
+    // needed to stop compilation
+    report_fatal_error("Invalid Root Signature Definition", false);
+    return std::nullopt;
+  }
+
+  return MRS;
+}
+
+AnalysisKey RootSignatureAnalysis::Key;
+
+std::optional<ModuleRootSignature>
+RootSignatureAnalysis::run(Module &M, ModuleAnalysisManager &AM) {
+  ModuleMetadataInfo MMI = AM.getResult<DXILMetadataAnalysis>(M);
+  if (MMI.ShaderProfile == Triple::Library)
+    return std::nullopt;
+  return ModuleRootSignature::analyzeModule(M, getEntryFunction(M, MMI));
+}
+
+//===----------------------------------------------------------------------===//
+bool RootSignatureAnalysisWrapper::runOnModule(Module &M) {
+  dxil::ModuleMetadataInfo &MMI =
+      getAnalysis<DXILMetadataAnalysisWrapperPass>().getModuleMetadata();
+  if (MMI.ShaderProfile == Triple::Library)
+    return false;
+  MRS = ModuleRootSignature::analyzeModule(M, getEntryFunction(M, MMI));
+  return false;
+}
+
+void RootSignatureAnalysisWrapper::getAnalysisUsage(AnalysisUsage &AU) const {
+  AU.setPreservesAll();
+  AU.addRequired<DXILMetadataAnalysisWrapperPass>();
+}
+
+char RootSignatureAnalysisWrapper::ID = 0;
+
+INITIALIZE_PASS_BEGIN(RootSignatureAnalysisWrapper,
+                      "dx-root-signature-analysis",
```

Comment:
This should be called `dxil-root-signature-analysis`. The `dx-` prefix seems to be used in DXILShaderFlags for some reason, but generally we've been consistent with the dxil- prefix.